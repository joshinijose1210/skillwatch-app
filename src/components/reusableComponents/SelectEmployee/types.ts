import React from 'react';

export interface SearchEmployeeProps {
    value: string | undefined;
    onChange: React.Dispatch<React.SetStateAction<string>>;
    disabled?: boolean;
    placeholder?: string;
    label?: string;
}
