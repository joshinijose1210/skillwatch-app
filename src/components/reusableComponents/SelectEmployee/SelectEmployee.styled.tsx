import styled from 'styled-components';

export const SelectEmployeeWrapper = styled.div`
    min-width: 500px;
`;
