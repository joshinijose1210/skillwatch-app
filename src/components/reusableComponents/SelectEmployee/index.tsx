export * from './SelectEmployee';
export * from './useSelectEmployee';
