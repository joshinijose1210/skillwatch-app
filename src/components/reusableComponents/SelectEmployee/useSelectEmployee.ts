import { useAppSelector } from '@slice';
import { useGetEmployeeListQuery } from '@slice/services';
import { useCallback } from 'react';
import { useLocation } from 'react-router-dom';

export const useSelectEmployee = () => {
    const userDetails = useAppSelector(state => state.user);
    const path = useLocation().pathname;

    const { data: employees, isSuccess: isEmployeeListReceived } = useGetEmployeeListQuery({
        path: '',
        params: [
            { name: 'teamId', value: -99 },
            { name: 'designationId', value: -99 },
            { name: 'organisationId', value: userDetails.organisationId },
            { name: 'roleId', value: -99 },
            { name: 'sortOrderId', value: 1 },
            { name: 'departmentId', value: -99 }
        ]
    });
    const employeeList = useCallback(() => {
        const filteredList = isEmployeeListReceived ? employees?.employees.filter((employee: any) => employee.status === true) : [];
        let list = filteredList
            ? filteredList.map((employee: any) => {
                  return {
                      value: employee.id,
                      label: `${employee.firstName} ${employee.lastName} (${employee.employeeId})`
                  };
              })
            : [];

        list = path.includes('dashboard/view-appreciation') ? list : list.filter((employee: any) => employee.value !== userDetails.id);
        list.sort((a: any, b: any) => a.label.localeCompare(b.label));
        return list;
    }, [isEmployeeListReceived, employees?.employees, path, userDetails.id]);

    return { employeeList };
};
