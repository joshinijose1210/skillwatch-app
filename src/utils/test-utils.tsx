import { ToastContainer } from '@medly-components/core';
import { setupListeners } from '@reduxjs/toolkit/dist/query';
import { store as rtkStore, setupStore } from '@slice';
import { RenderOptions, RenderResult, render } from '@testing-library/react';
import { defaultTheme } from '@theme';
import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import reduxMockStore from 'redux-mock-store';
import { ThemeProvider } from 'styled-components';
import { userMockState } from '../mocks/userMockState';
import { initialState } from '../slice';

interface CustomOptions extends RenderOptions {
    initialState?: any;
    initialEntries?: any;
}

interface WithThemeProviderProps {
    children?: React.ReactNode;
}

const mockStore = reduxMockStore(),
    store = mockStore({
        ...initialState,
        user: { ...initialState.user, groups: ['admin'] }
    });

const WithThemeProvider: React.FunctionComponent<WithThemeProviderProps> = ({ children }) => (
    <ThemeProvider theme={defaultTheme}>
        <>
            <ToastContainer position="top-end" />
            {children}
        </>
    </ThemeProvider>
);

const WithStore: React.FunctionComponent<WithThemeProviderProps> = ({ children }) => (
    <Provider store={store}>
        <WithThemeProvider>{children}</WithThemeProvider>
    </Provider>
);

const WithRouter: React.FunctionComponent<WithThemeProviderProps> = ({ children }) => (
    <MemoryRouter>
        <WithThemeProvider>{children}</WithThemeProvider>
    </MemoryRouter>
);

const WithStoreAndRouter: React.FunctionComponent<WithThemeProviderProps> = ({ children }) => (
    <Provider store={store}>
        <MemoryRouter>
            <WithThemeProvider>{children}</WithThemeProvider>
        </MemoryRouter>
    </Provider>
);

export const RTKRender = (ui: React.ReactElement, { initialState = userMockState, ...renderOptions }: CustomOptions = {}) => {
    const store = setupStore(initialState);
    setupListeners(store.dispatch);

    const RtkWrapper: React.FunctionComponent<WithThemeProviderProps> = ({ children }) => (
        <Provider store={store}>
            <MemoryRouter initialEntries={renderOptions.initialEntries}>
                <WithThemeProvider>{children}</WithThemeProvider>
            </MemoryRouter>
        </Provider>
    );
    return { rtkStore, ...render(ui, { wrapper: RtkWrapper, ...renderOptions }) };
};
const customRender = (ui: React.ReactElement, options?: RenderOptions): RenderResult =>
    render(ui, { wrapper: WithThemeProvider, ...options });

export const renderWithStore = (ui: React.ReactElement, options?: RenderOptions): RenderResult =>
    render(ui, { wrapper: WithStore, ...options });

export const renderWithRouter = (ui: React.ReactElement, options?: RenderOptions): RenderResult =>
    render(ui, { wrapper: WithRouter, ...options });

export const renderWithStoreAndRouter = (ui: React.ReactElement, options?: RenderOptions): RenderResult =>
    render(ui, { wrapper: WithStoreAndRouter, ...options });

export const wrapper = (props: { children: React.ReactNode }, { initialState = userMockState }: CustomOptions = {}) => {
    const store = setupStore(initialState);
    setupListeners(store.dispatch);
    return (
        <Provider store={store}>
            <MemoryRouter>
                <WithThemeProvider>{props.children}</WithThemeProvider>
            </MemoryRouter>
        </Provider>
    );
};

// re-export everything
export * from '@testing-library/react';
// override render method
export { customRender as render, store };
