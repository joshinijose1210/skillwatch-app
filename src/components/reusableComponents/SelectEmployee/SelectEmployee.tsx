import { StyledSingleSelect } from '@common';
import * as Styled from './SelectEmployee.styled';
import { SearchEmployeeProps } from './types';
import { useSelectEmployee } from './useSelectEmployee';
import TooltipDropdown from '@components/reusableComponents/TooltipDropdown';
import { OptionsType } from '@components/reusableComponents/TooltipDropdown/types';

export const SelectEmployee = ({
    value,
    onChange,
    disabled = false,
    placeholder = 'Select Employee Name',
    label = 'Employee Name'
}: SearchEmployeeProps) => {
    const { employeeList } = useSelectEmployee();

    return (
        <Styled.SelectEmployeeWrapper>
            <StyledSingleSelect
                disabled={disabled}
                options={employeeList()}
                value={value}
                onChange={val => val && onChange(val)}
                variant="outlined"
                isSearchable={true}
                size="M"
                placeholder={placeholder}
                label={label}
                minWidth="100%"
            />
            <TooltipDropdown
                dataIds={[`${label.toLowerCase()}-input`]}
                values={value && employeeList().length ? employeeList().filter((item: OptionsType) => value === item.value) : []}
            />
        </Styled.SelectEmployeeWrapper>
    );
};
