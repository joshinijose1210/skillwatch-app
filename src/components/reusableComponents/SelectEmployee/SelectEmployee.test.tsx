import { RTKRender } from '@utils/test-utils';
import { SelectEmployee } from './SelectEmployee';

describe('SearchEmployee component', () => {
    it('should take snapshot', () => {
        const { container } = RTKRender(<SelectEmployee value={''} onChange={jest.fn()} />);
        expect(container).toMatchSnapshot();
    });
});
