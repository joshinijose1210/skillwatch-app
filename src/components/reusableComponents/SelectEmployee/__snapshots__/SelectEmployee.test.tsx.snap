// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchEmployee component should take snapshot 1`] = `
<div>
  <div
    class="sc-tpub7j-0 iWOTdT"
  >
    <div
      class="sc-1hcsztp-0 cwjrcY sc-tkiih6-13 cecHDR"
      id="employee name-wrapper"
      minwidth="100%"
      variant="outlined"
    >
      <div
        class="sc-17d4s39-0 khebHA"
        id="employee name-input-wrapper"
        minwidth="100%"
      >
        <div
          class="sc-19px6zp-0 BrSGZ"
          filled="[object Object]"
          fusion="[object Object]"
          minrows="1"
          outlined="[object Object]"
          textvariant="[object Object]"
          variant="outlined"
        >
          <div
            class="sc-lhr232-0 btNOKw"
            variant="outlined"
          >
            <input
              aria-describedby="employee name-helper-text"
              autocomplete="off"
              class="sc-hb99wz-0 jaEbhp"
              data-tooltip-id="employee name-input"
              errortext=""
              id="employee name-input"
              inputsize="M"
              placeholder="Select Employee Name"
              type="text"
              value=""
              variant="outlined"
            />
            <label
              class="sc-1du51u0-0 eMZrST"
              for="employee name-input"
              inputwidth="0"
              variant="outlined"
            >
              Employee Name
            </label>
          </div>
          <svg
            class="sc-1pf3x8a-0 klDMsE sc-dynlvq-0 bdmviR"
            fill="none"
            height="1em"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12.792 14.97a1 1 0 0 1-1.584 0l-3.165-4.11a1 1 0 0 1 .793-1.61h6.328a1 1 0 0 1 .793 1.61l-3.165 4.11z"
              fill="#000"
            />
          </svg>
        </div>
        
      </div>
    </div>
    
  </div>
</div>
`;
